agent/service_manager.go

```diff
@@ -379,39 +379,76 @@ func mergeServiceConfig(defaults *structs.ServiceConfigResponse, service *struct
 	if ns.Proxy.MeshGateway.Mode == structs.MeshGatewayModeDefault {
 		ns.Proxy.MeshGateway.Mode = defaults.MeshGateway.Mode
 	}
+	if !ns.Proxy.TransparentProxy {
+		ns.Proxy.TransparentProxy = defaults.TransparentProxy
+	}
+
+	// seenUpstreams stores the upstreams seen from the local registration so that we can also add synthetic entries.
+	// for upstream configuration that was defined via service-defaults.UpstreamConfigs. In TransparentProxy mode
+	// ns.Proxy.Upstreams will likely be empty because users do not need to define upstreams explicitly.
+	// So to store upstream-specific flags from central config, we add entries to ns.Proxy.Upstream with thosee values.
+	remoteUpstreams := make(map[structs.ServiceID]structs.Upstream)
+
+	for _, us := range defaults.UpstreamIDConfigs {
+		parsed, err := structs.ParseUpstreamConfig(us.Config)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse upstream config map for %s: %v", us.Upstream.String(), err)
+		}
 
-	// Merge upstream defaults if there were any returned
+		// Delete the mesh gateway key since this is the only place it is read from an opaque map.
+		// Later reads use Proxy.MeshGateway.
+		delete(us.Config, "mesh_gateway")
+
+		remoteUpstreams[us.Upstream] = structs.Upstream{
+			DestinationNamespace: us.Upstream.NamespaceOrDefault(),
+			DestinationName:      us.Upstream.ID,
+			Config:               us.Config,
+			MeshGateway:          parsed.MeshGateway,
```

Comment:
Did you mean to lift the transparentproxy flag here as well?

Reply:
Nope, the TransparentProxy flag on an upstream definition won't do anything. It only kicks in as a property of the downstream proxy's registration.